router.go

```diff
@@ -63,39 +72,76 @@ func (m MapRouter) Register(rs []transport.Procedure) {
 			panic("Expected procedure name not to be empty string in registration")
 		}
 
+		if r.Encoding != "" {
+			// Route to individual handlers for unique combinations of service,
+			// procedure, and encoding. This shall henceforth be the
+			// recommended way for models to register procedures.
+			spe := serviceProcedureEncoding{
+				service:   r.Service,
+				procedure: r.Name,
+				encoding:  r.Encoding,
+			}
+			m.serviceProcedureEncodings[spe] = r
+		}
+
+		// This supports wild card encodings (for backward compatibility,
+		// since type models like Thrift were not previously required to
+		// specify the encoding of every procedure)
 		sp := serviceProcedure{
 			service:   r.Service,
 			procedure: r.Name,
 		}
-		m.entries[sp] = r
+		// Protect against registering ambiguous wildcards
+		if _, ok := m.serviceProcedures[sp]; r.Encoding == "" && ok {
+			panic(fmt.Sprintf("Cannot register multiple procedures with the same name and service without specifying alternate encodings, got ambiguous procedures for (%q, %q)", sp.service, sp.procedure))
+		}
+		m.serviceProcedures[sp] = r
+
 	}
 }
 
 // Procedures returns a list procedures that
 // have been registered so far.
 func (m MapRouter) Procedures() []transport.Procedure {
-	procs := make([]transport.Procedure, 0, len(m.entries))
-	for _, v := range m.entries {
+	procs := make([]transport.Procedure, 0, len(m.serviceProcedures)+len(m.serviceProcedureEncodings))
+	for _, v := range m.serviceProcedureEncodings {
+		procs = append(procs, v)
+	}
+	for _, v := range m.serviceProcedures {
```

Comment:
Addressed in #1053 which makes (service, procedure) and (service, procedure, encoding) routing tables mutually exclusive.